File: 프로그래머스/2/42586. 기능개발/기능개발.java
import java.util.*;

class Solution {
    public int[] solution(int[] progresses, int[] speeds) {
        Queue<Integer> q = new LinkedList<>();
        for(int i=0; i < speeds.length; i++){
            int remainDays = 100 - progresses[i];
            int days = (int) Math.ceil( (double) remainDays / speeds[i]);
            q.add(days);
        }
        
        List<Integer> list = new ArrayList<>();
        while(!q.isEmpty()){
            int currentDay = q.poll();
            int count = 1;
            while(!q.isEmpty() && q.peek() <= currentDay){
                q.poll();
                count++;
            }
            
            list.add(count);
        }
        return list.stream().mapToInt(i->i).toArray();
    }
}

File: 프로그래머스/2/42586. 기능개발/기능개발.js
class Queue {
    constructor(initial = []) {
        this.items = Array.isArray(initial) ? [...initial] : [];
        this.front = 0;
        this.rear = this.items.length; 
    }

    enqueue(item) {
        this.items.push(item);
        this.rear++;
    }

    dequeue() {
        const value = this.items[this.front];
        delete this.items[this.front++];
        return value;
    }
    
    peek() {
        return this.items[this.front];
    }

    size() {
        return this.rear - this.front;
    }
}
            
function solution(progresses, speeds) {
    const queue = new Queue(progresses);
    let count = 1, tempCount = 0, prevValue = 0;
    const answer = [];
    
    for(let i = 0; i < speeds.length; i++){
        let value = queue.peek() + speeds[i] * count;
        while(value < 100){ 
            count++;    
            value = queue.peek() + speeds[i] * count; 
        }
        queue.dequeue();
        if(prevValue === 0 || count === prevValue){
            tempCount++;
        } else {
            answer.push(tempCount);
            tempCount = 1;
        }
        prevValue = count;
    }
    answer.push(tempCount);
    return answer;
}